### ISSUE TITLE
Unreported accidental dead code when using `isinstance` 

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#5393]
**Describe the bug**
Using `if isinstance(some_union, TypeNotInUnion): ...` produces a branch of dead code that is not flagged by pyright, which could result in bugs

**To Reproduce**
```python
# pyright: strict

def could_fail() -> "float | None":
    return 3.14

def x() -> float:
    result = could_fail()
    if isinstance(result, int): # Oops. Should've checked for float, but no warnings
        return result # no errors reported here because `result` is of type `Never`
    return 123.45
```

**Expected behavior**
I would expect an error on `if isinstance(result, str):`, alerting the user that the condition will always evaluate to `False`.

**VS Code extension or command-line**
extension `v1.1.315`

**Comment by GugelRobin** [comment]
Hello @Tomaz-Vieira
This isn't a bug - as far as I know this is as designed.
I'd classify this as a feature-request to add a `reportUnreachableCode` flag which would (possibly) default to `error` on strict mode.

Sadly, at least after trying for like 20 minutes, I haven't found any way to raise an error when a value is inferred to be `Never` - there's only an option to raise an error if a value is _expected_ to be `Never` but isn't.

**Comment by debonte** [comment]
Given that `int` is a subtype of `float`, I assume that in your sample code above you meant to say `isinstance(result, str)` instead of `isinstance(result, int)`.

In that case, I'm not seeing `result` reported as type `Never` in the `if` block, but rather `<subclass of float and str>`. So my guess is that Pyright doesn't report an error here because in the general case (ex. with multiple inheritance) it's possible that the value returned by `could_fail` could match both the return type of `could_fail` and the test type passed to `isinstance`.

In this specific case (`float` and `str`) that doesn't appear to be supported at runtime. The following class gives me `TypeError
multiple bases have instance lay-out conflict`, but Pyright is not aware of that conflict:

```python
class MyFloatStr(float, str):
    pass
```

But consider this example:

```python
class A: pass
class B: pass
class Both(A, B): pass

def could_fail() -> A | None:
    return Both()

def x() -> A:
    result = could_fail()
    if isinstance(result, B): 
        return result 
    return A()
```

**Comment by Tomaz-Vieira** [comment]
I could swear I saw the `Never` type show up when I was writing my actual code, but I can't seem to reproduce it again; It might have been some stale incremental-checking cache or something.

Those are great points, @debonte, thanks. I guess I have to rethink the way I look at `Union`s now.
